core/src/main/java/com/scalar/db/api/ScanBuilder.java

```diff
@@ -548,4 +847,277 @@ public Scan build() {
+
+    private BuildableScanAllFromExistingWithOngoingWhereOr(
+        BuildableScanAllFromExistingWithOngoingWhere buildable) {
+      super(buildable);
+    }
+
+    private BuildableScanAllFromExistingWithOngoingWhereOr(
+        BuildableScanOrScanAllFromExisting buildable, AndConditionSet andConditionSet) {
+      super(buildable);
+      conjunctions.add(Scan.Conjunction.of(andConditionSet.getConditions()));
+    }
+
+    @Override
+    public BuildableScanAllFromExistingWithOngoingWhereOr or(ConditionalExpression condition) {
+      checkNotNull(condition);
+      conjunctions.add(Scan.Conjunction.of(condition));
+      return this;
+    }
+
+    @Override
+    public BuildableScanAllFromExistingWithOngoingWhereOr or(AndConditionSet andConditionSet) {
+      checkNotNull(andConditionSet);
+      conjunctions.add(Scan.Conjunction.of(andConditionSet.getConditions()));
+      return this;
+    }
+  }
+
+  public static class BuildableScanAllFromExistingWithOngoingWhereAnd
+      extends BuildableScanAllFromExistingWithOngoingWhere
+      implements And<BuildableScanAllFromExistingWithOngoingWhereAnd> {
+
+    private BuildableScanAllFromExistingWithOngoingWhereAnd(
+        BuildableScanAllFromExistingWithOngoingWhere buildable) {
+      super(buildable);
+    }
+
+    private BuildableScanAllFromExistingWithOngoingWhereAnd(
+        BuildableScanOrScanAllFromExisting buildable, OrConditionSet orConditionSet) {
+      super(buildable);
+      disjunctions.add(orConditionSet.getConditions());
+    }
+
+    @Override
+    public BuildableScanAllFromExistingWithOngoingWhereAnd and(ConditionalExpression condition) {
+      checkNotNull(condition);
+      disjunctions.add(ImmutableSet.of(condition));
+      return this;
+    }
+
+    @Override
+    public BuildableScanAllFromExistingWithOngoingWhereAnd and(OrConditionSet orConditionSet) {
+      checkNotNull(orConditionSet);
+      disjunctions.add(orConditionSet.getConditions());
+      return this;
+    }
+  }
+
+  private static Scan buildScanAllWithConjunctions(
+      String namespaceName,
+      String tableName,
+      @Nullable ConditionalExpression condition,
+      Set<Scan.Conjunction> conjunctions,
+      List<Set<ConditionalExpression>> disjunctions,
+      List<String> projections,
+      List<Scan.Ordering> orderings,
+      int limit,
+      @Nullable com.scalar.db.api.Consistency consistency) {
+    ScanAll scan = new ScanAll();
+    scan.forNamespace(namespaceName).forTable(tableName).withLimit(limit);
+    orderings.forEach(scan::withOrdering);
+
+    if (condition != null) {
+      assert conjunctions.isEmpty() && disjunctions.isEmpty();
+      scan.withConjunctions(ImmutableSet.of(Scan.Conjunction.of(condition)));
+    } else if (conjunctions.isEmpty()) {
+      scan.withConjunctions(
+          Sets.cartesianProduct(disjunctions).stream()
+              .map(Scan.Conjunction::of)
+              .collect(Collectors.toSet()));
```

Comment:
When building a where clause with ".and().and()...", the builder holds a (conjunctive) set of disjunctions, and its conjunction should be empty. In this case, we convert CNF to DNF according to the distributive law.